common/repeat_timer.go

```diff
@@ -5,152 +5,224 @@ import (
 	"time"
 )
 
+// Used by RepeatTimer the first time,
+// and every time it's Reset() after Stop().
+type TickerMaker func(dur time.Duration) Ticker
+
 // Ticker is a basic ticker interface.
 type Ticker interface {
+
+	// Never changes, never closes.
 	Chan() <-chan time.Time
+
+	// Stopping a stopped Ticker will panic.
 	Stop()
-	Reset()
 }
 
-// DefaultTicker wraps the stdlibs Ticker implementation.
-type DefaultTicker struct {
-	t   *time.Ticker
-	dur time.Duration
-}
+//----------------------------------------
+// defaultTickerMaker
 
-// NewDefaultTicker returns a new DefaultTicker
-func NewDefaultTicker(dur time.Duration) *DefaultTicker {
-	return &DefaultTicker{
-		time.NewTicker(dur),
-		dur,
-	}
+func defaultTickerMaker(dur time.Duration) Ticker {
+	ticker := time.NewTicker(dur)
+	return (*defaultTicker)(ticker)
 }
 
+type defaultTicker time.Ticker
+
 // Implements Ticker
-func (t *DefaultTicker) Chan() <-chan time.Time {
-	return t.t.C
+func (t *defaultTicker) Chan() <-chan time.Time {
+	return t.C
 }
 
 // Implements Ticker
-func (t *DefaultTicker) Stop() {
-	t.t.Stop()
-	t.t = nil
+func (t *defaultTicker) Stop() {
+	t.Stop()
 }
 
-// Implements Ticker
-func (t *DefaultTicker) Reset() {
-	t.t = time.NewTicker(t.dur)
+//----------------------------------------
+// LogicalTickerMaker
+
+// Construct a TickerMaker that always uses `ch`.
+// It's useful for simulating a deterministic clock.
+func NewLogicalTickerMaker(ch chan time.Time) TickerMaker {
+	return func(dur time.Duration) Ticker {
+		return newLogicalTicker(ch, dur)
+	}
 }
 
-// ManualTicker wraps a channel that can be manually sent on
-type ManualTicker struct {
-	ch chan time.Time
+type logicalTicker struct {
+	source <-chan time.Time
+	ch     chan time.Time
+	quit   chan struct{}
 }
 
-// NewManualTicker returns a new ManualTicker
-func NewManualTicker(ch chan time.Time) *ManualTicker {
-	return &ManualTicker{
-		ch: ch,
+func newLogicalTicker(source <-chan time.Time, interval time.Duration) Ticker {
+	lt := &logicalTicker{
+		source: source,
+		ch:     make(chan time.Time),
+		quit:   make(chan struct{}),
 	}
+	go lt.fireRoutine(interval)
+	return lt
 }
 
-// Implements Ticker
-func (t *ManualTicker) Chan() <-chan time.Time {
-	return t.ch
```

Comment:
Since t.ch was being used as both source and `Chan()` (emitter), there was no need for another goroutine.  But the implementation is broken because a Ticker is not identical to the source.  The source is a background ticker that provides a continuous heartbeat to the program (duplicated in a for-loop for each waiting go-routine), not a ticker with a custom interval.  So we need to split this into two channels and connect them with a go-routine that filters source events before passing them on to the `.Chan()` emitter.